zcash_client_backend/src/fees/zip317.rs

```diff
@@ -184,6 +288,132 @@ mod tests {
+                    None,
+                    Some(&WalletMeta::new(
+                        existing_notes,
+                        #[cfg(feature = "orchard")]
+                        0,
+                    )),
+                )
+            };
+
+            assert_matches!(
+                balance(0),
+                Ok(balance) if
+                    balance.proposed_change() == [
+                        ChangeValue::sapling(NonNegativeAmount::const_from_u64(129_4000), None),
+                        ChangeValue::sapling(NonNegativeAmount::const_from_u64(129_4000), None),
+                        ChangeValue::sapling(NonNegativeAmount::const_from_u64(129_4000), None),
+                        ChangeValue::sapling(NonNegativeAmount::const_from_u64(129_4000), None),
+                        ChangeValue::sapling(NonNegativeAmount::const_from_u64(129_4000), None),
+                    ] &&
+                    balance.fee_required() == NonNegativeAmount::const_from_u64(30000)
+            );
+
+            assert_matches!(
+                balance(2),
+                Ok(balance) if
+                    balance.proposed_change() == [
+                        ChangeValue::sapling(NonNegativeAmount::const_from_u64(216_0000), None),
+                        ChangeValue::sapling(NonNegativeAmount::const_from_u64(216_0000), None),
+                        ChangeValue::sapling(NonNegativeAmount::const_from_u64(216_0000), None),
+                    ] &&
+                    balance.fee_required() == NonNegativeAmount::const_from_u64(20000)
+            );
+        }
+
+        {
+            // spend a single Sapling note and produce 4 outputs, as the value of the note isn't
+            // sufficient to produce 5
+            let result = change_strategy.compute_balance(
+                &Network::TestNetwork,
+                Network::TestNetwork
+                    .activation_height(NetworkUpgrade::Nu5)
+                    .unwrap(),
+                &[] as &[TestTransparentInput],
+                &[] as &[TxOut],
+                &(
+                    sapling::builder::BundleType::DEFAULT,
+                    &[TestSaplingInput {
+                        note_id: 0,
+                        value: NonNegativeAmount::const_from_u64(600_0000),
+                    }][..],
+                    &[SaplingPayment::new(NonNegativeAmount::const_from_u64(
+                        100_0000,
+                    ))][..],
+                ),
+                #[cfg(feature = "orchard")]
+                &orchard_fees::EmptyBundleView,
+                None,
+                Some(&WalletMeta::new(
+                    0,
+                    #[cfg(feature = "orchard")]
+                    0,
+                )),
+            );
+
+            // FIXME: The current splitting strategy may result in overpaying fees by a small amount.
+            // This may be acceptable for an initial implementation.
```

Comment:
To fix this, can we recompute the fee once we have working outputs, and then add the excess to the first change output? That seems like a simple way to address this inside the transaction builder rather than in this fee calculator. I don't think we need to aim for perfectly-split notes (the first note already differs due to the remainder).

Reply:
Yup, I'm adding a commit to fix this.